### PR TITLE
Add power battery% access

### PR DESCRIPTION
**Pull request by StraToN**
(repost of #5869 which was broken)
Only tested X11 and Windows by @vnen. Sorry I don't have a mac.

JOB IS NOT FINISHED FOR ALL PLATFORMS! DON'T MERGE BEFORE ALL PLATFORMS ARE TESTED !!

NEED TESTERS for iOS, OSX, WinRT and Android (may not compile for these).
Thanks

### CODE REVIEW

**Comment by vnen** [comment]
BTW, I wouldn't bother (for now) about WinRT. Right now it does not even compile.

File: platform/android/os_android.h [hunk 1]
```diff
@@ -154,6 +155,8 @@ class OS_Android : public OS_Unix {
 	VideoPauseFunc video_pause_func;
 	VideoStopFunc video_stop_func;
 	SetKeepScreenOnFunc set_keep_screen_on_func;
+	
+	power_android power_manager;
```
Comment (vnen) [inline platform/android/os_android.h]: Should be `power_android *power_manager;`, right? This is what breaks the build for Android.

Comment (StraToN) [inline platform/android/os_android.h]: That's corrected. I just pushed the fix, but I can't affirm android will build, I'm almost certain there are other problems, especially in ``platform/android/power_android.cpp``'s method ``int Android_JNI_GetPowerInfo(int* plugged, int* charged, int* battery, int* seconds, int* percent)``

**Comment by akien-mga** [comment]
It fails building on OSX and Android: https://travis-ci.org/godotengine/godot/builds/146963962

**Comment by StraToN** [comment]
@akien-mga Yes, I don't have a Mac to test, nor I have an Android environment.

**Comment by akien-mga** [comment]
Yes but you can check the build logs and see what are the issues :)

For Android you'd likely have to setup a build environment to test properly though, as there seem to be various issues: https://travis-ci.org/godotengine/godot/jobs/146963965
For OSX it looks like you just forgot to translate a SDL2 type to Dictionary: https://travis-ci.org/godotengine/godot/jobs/146963967

**Comment by GungnirInd** [comment]
Adding ```#include <CoreFoundation/CoreFoundation.h>``` to ```platform/osx/power_osx.h``` and ```'-framework', 'IOKit'``` to the LDFLAGS in ```platform/osx/detect.py``` gets it to compile and work fine on OS X.  I haven't tested iOS, but it seems to compile.

**Comment by StraToN** [comment]
@vnen I think it's possible to implement WinRT pretty easily using the Battery class documented here http://msdn.microsoft.com/library/windowsphone/develop/jj207231, according to SDL comment left in `` platform/winrt/power_winrt.cpp:41``. I didn't take the time to do it as I couldn't test it anyway, however it seems we only need to return the values obtained with ``Battery::GetDefault().RemainingChargePercent`` and ``Battery::GetDefault().RemainingDischargeTime``.

@GungnirInd Thanks a lot, it worked :) 
Now testing is needed for OSX and iOS, for people insterested, thanks in advance!

**Comment by vnen** [comment]
I need to test this for Android. It'd be nice to merge this before 2.2. Since now I'm fixing the WinRT port, it can be added too for such platform.

**Comment by StraToN** [comment]
@vnen I'll try to get it to compile with Android now, even though I am not an Android developer. I suppose some includes are missing.

**Comment by reduz** [comment]
What's the status of this one?

**Comment by StraToN** [comment]
@reduz Still missing Android development by converting SDL code to Godot. I am not competent on this one yet 😕

**Comment by reduz** [comment]
you can ask @RandomShaper for help I guess

**Comment by akien-mga** [comment]
@StraToN The PR has conflicts, could you have a look at rebasing?

**Comment by StraToN** [comment]
@akien-mga Will do ASAP.

**Comment by StraToN** [comment]
Rebased. I suspect the branch is now ugly, still it needs some testing anyway.

**Comment by StraToN** [comment]
Rebased.
Thanks again to @akien-mga for the (now usual) help :)

**Comment by akien-mga** [comment]
Looking at the commit history, there's still something fishy though.. :/

**Comment by StraToN** [comment]
Indeed, looks like I left 2 useless commits . Will correct it tonight

**Comment by StraToN** [comment]
@akien-mga Corrected. Do you validate? :)

**Comment by akien-mga** [comment]
Looks good commit-wise from a quick glance :)

File: platform/winrt/power_winrt.cpp [hunk 2]
```diff
@@ -0,0 +1,78 @@
+/*************************************************************************/
+/*  power_winrt.h                                                   */
```
Comment (vnen) [inline platform/winrt/power_winrt.cpp]: Forgot to change the file name here.

Comment (StraToN) [inline platform/winrt/power_winrt.cpp]: Corrected. Thanks !

**Comment by akien-mga** [comment]
@StraToN You have some rebase work ahead :)

**Comment by StraToN** [comment]
@akien-mga Done ;)

File: platform/osx/power_osx.h [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+#include <CoreFoundation/CoreFoundation.h>
+
+class power_osx {
+
+private:
+	int nsecs_left;
+	int percent_left;
+	PowerState power_state;
+	void checkps(CFDictionaryRef dict, bool * have_ac, bool * have_battery, bool * charging);
+	bool GetPowerInfo_MacOSX(/*PowerState * state, int *seconds, int *percent*/);
```
Comment (StraToN) [inline platform/osx/power_osx.h]: I should kick this comment out.

**Comment by akien-mga** [comment]
Merged manually in 94103c0.

Please open a new issue to keep track of the state of this feature on all platforms (implemented/working) so that we make sure it's all good for 3.0.

**Pull request closed without merging**